File: frontend/src/lib/platforms/youtubeYtdlp.ts
import { execFile } from "node:child_process";
import { promisify } from "node:util";

import type { CandidateLink, ProductInput, SearchSettings } from "../types";
import { primaryQuery } from "../scoring";

const execFileAsync = promisify(execFile);

function parsePlaylistJson(raw: string): CandidateLink[] {
  const info = JSON.parse(raw) as {
    _type?: string;
    entries?: unknown[];
    id?: string;
    title?: string;
    url?: string;
    description?: string;
  };
  let entries: unknown[];
  if (info._type === "playlist") {
    entries = (info.entries ?? []).filter(Boolean);
  } else if (info.id || info.url) {
    entries = [info];
  } else {
    entries = [];
  }
  const out: CandidateLink[] = [];
  for (const ent of entries) {
    if (!ent || typeof ent !== "object") continue;
    const e = ent as Record<string, unknown>;
    const id = typeof e.id === "string" ? e.id : "";
    const title = String(e.title ?? "");
    let url = String(e.url ?? "");
    if (!url && id) url = `https://www.youtube.com/watch?v=${id}`;
    if (!url) continue;
    out.push({
      media: "Youtube",
      brand: "", // filled by caller
      url,
      sku: "",
      productName: "",
      title,
      snippet: String(e.description ?? "").slice(0, 500),
      score: 0,
    });
  }
  return out;
}

async function runYtdlpJson(args: string[]): Promise<string | null> {
  const attempts: { cmd: string; argv: string[] }[] = [
    { cmd: process.env.YTDLP_BIN || "yt-dlp", argv: args },
    { cmd: "python", argv: ["-m", "yt_dlp", ...args] },
    { cmd: "py", argv: ["-m", "yt_dlp", ...args] },
  ];
  for (const { cmd, argv } of attempts) {
    try {
      const { stdout } = await execFileAsync(cmd, argv, {
        timeout: 90_000,
        maxBuffer: 25 * 1024 * 1024,
        windowsHide: true,
      });
      return stdout.toString();
    } catch {
      /* try next */
    }
  }
  return null;
}

/**
 * YouTube search via yt-dlp (Innertube). Requires `yt-dlp` on PATH or Python with yt-dlp installed.
 */
export async function searchYoutubeYtdlp(
  settings: SearchSettings,
  product: ProductInput,
  queries: string[]
): Promise<CandidateLink[]> {
  if (!settings.youtubeUseYtdlp) return [];
  const cap = settings.maxResultsPerPlatform;
  const qText = (queries.length
    ? queries.slice(0, 3).join(" ")
    : primaryQuery(product)
  )
    .trim()
    .slice(0, 200);
  if (!qText) return [];

  const args = [
    `ytsearch${cap}:${qText}`,
    "-J",
    "--flat-playlist",
    "--no-download",
    "--no-warnings",
    "--quiet",
  ];
  const stdout = await runYtdlpJson(args);
  if (!stdout) return [];

  try {
    const rows = parsePlaylistJson(stdout);
    const brand = product.brand.trim();
    const sku = product.sku.trim();
    const productName = product.productName.trim();
    return rows.map((r) => ({
      ...r,
      brand,
      sku,
      productName,
      sourceQuery: qText,
    }));
  } catch {
    return [];
  }
}

File: src/linksearch/platforms/youtube_ytdlp.py
from __future__ import annotations

import logging
from typing import TYPE_CHECKING

from linksearch.models import CandidateLink, ProductInput

if TYPE_CHECKING:
    from linksearch.config import Settings

logger = logging.getLogger(__name__)


def search_youtube_ytdlp_sync(
    settings: Settings,
    product: ProductInput,
    queries: list[str],
) -> list[CandidateLink]:
    """YouTube search via yt-dlp (Innertube). No Google Cloud API key."""
    if not settings.youtube_use_ytdlp:
        return []
    try:
        import yt_dlp
    except ImportError:
        logger.warning("yt-dlp is not installed; skip Innertube YouTube search.")
        return []

    cap = max(1, settings.max_results_per_platform)
    q_text = " ".join(queries[:3]) if queries else product.primary_query()
    q_text = q_text.strip()[:200]
    if not q_text:
        return []

    ydl_opts: dict = {
        "quiet": True,
        "no_warnings": True,
        "extract_flat": "in_playlist",
        "skip_download": True,
        "socket_timeout": min(60.0, max(15.0, settings.request_timeout_seconds + 15)),
    }
    url = f"ytsearch{cap}:{q_text}"
    try:
        with yt_dlp.YoutubeDL(ydl_opts) as ydl:
            info = ydl.extract_info(url, download=False)
    except Exception as e:
        logger.warning("yt-dlp YouTube search failed: %s", e)
        return []

    entries: list = []
    if info is None:
        return []
    if info.get("_type") == "playlist":
        entries = [e for e in (info.get("entries") or []) if e]
    else:
        entries = [info]

    out: list[CandidateLink] = []
    seen: set[str] = set()
    brand = product.normalized_brand()
    sku = product.sku.strip()
    pname = product.product_name.strip()

    for ent in entries:
        if not isinstance(ent, dict):
            continue
        vid = ent.get("id")
        title = str(ent.get("title") or "")
        u = str(ent.get("url") or "")
        if not u and vid:
            u = f"https://www.youtube.com/watch?v={vid}"
        if not u:
            continue
        if u in seen:
            continue
        seen.add(u)
        out.append(
            CandidateLink(
                media="Youtube",
                brand=brand,
                url=u,
                sku=sku,
                product_name=pname,
                title=title,
                snippet=str(ent.get("description") or "")[:500],
                source_query=q_text,
            )
        )
        if len(out) >= cap:
            break
    return out
